Mathlib/AlgebraicGeometry/EllipticCurve/DivisionPolynomial/Basic.lean

```diff
@@ -0,0 +1,432 @@
+ * $\Phi_n = X^{n^2} + \dots$.
+
+## Main definitions
+
+ * `WeierstrassCurve.Ψ₂Sq`: the univariate polynomial $\Psi_2^{[2]}$.
+ * `WeierstrassCurve.preΨ`: the univariate polynomials $\tilde{\Psi}_n$.
+ * `WeierstrassCurve.ΨSq`: the univariate polynomials $\Psi_n^{[2]}$.
+ * `WeierstrassCurve.Ψ`: the bivariate polynomials $\Psi_n$.
+ * `WeierstrassCurve.Φ`: the univariate polynomials $\Phi_n$.
+ * `WeierstrassCurve.ψ`: the bivariate $n$-division polynomials $\psi_n$.
+ * `WeierstrassCurve.φ`: the bivariate polynomials $\phi_n$.
+ * TODO: the bivariate polynomials $\omega_n$.
+
+## Implementation notes
+
+Analogously to `Mathlib.NumberTheory.EllipticDivisibilitySequence`, the $n$-division bivariate
+polynomials $\Psi_n$ are defined in terms of the univariate polynomials $\tilde{\Psi}_n$. This is
+done partially to avoid ring division, but more crucially to allow the definition of $\Psi_n^{[2]}$
+and $\Phi_n$ as univariate polynomials without needing to work under the coordinate ring, and to
+allow the computation of their leading terms without ambiguity. Furthermore, evaluating these
+polynomials at a rational point on $W$ recovers their original definition up to linear combinations
+of the Weierstrass equation of $W$, hence also avoiding the need to work under the coordinate ring.
+
+TODO: implementation notes for the definition of $\omega_n$.
+
+## References
+
+[J Silverman, *The Arithmetic of Elliptic Curves*][silverman2009]
+
+## Tags
+
+elliptic curve, division polynomial, torsion point
+-/
+
+open Polynomial PolynomialPolynomial
+
+local macro "C_simp" : tactic =>
+  `(tactic| simp only [map_ofNat, C_0, C_1, C_neg, C_add, C_sub, C_mul, C_pow])
+
+universe u v
+
+namespace WeierstrassCurve
+
+variable {R : Type u} {S : Type v} [CommRing R] [CommRing S] (W : WeierstrassCurve R) (f : R →+* S)
+
+/-! ### The univariate polynomial $\Psi_2^{[2]}$ -/
+
+/-- The $2$-division polynomial $\psi_2 = \Psi_2$. -/
+protected noncomputable def ψ₂ : R[X][Y] :=
+  W.toAffine.polynomialY
+
+/-- The univariate polynomial $\Psi_2^{[2]}$ congruent to $\psi_2^2$. -/
+protected noncomputable def Ψ₂Sq : R[X] :=
+  C 4 * X ^ 3 + C W.b₂ * X ^ 2 + C (2 * W.b₄) * X + C W.b₆
+
+lemma C_Ψ₂Sq_eq : C W.Ψ₂Sq = W.ψ₂ ^ 2 - 4 * W.toAffine.polynomial := by
+  rw [WeierstrassCurve.Ψ₂Sq, WeierstrassCurve.ψ₂, b₂, b₄, b₆, Affine.polynomialY, Affine.polynomial]
+  C_simp
+  ring1
+
+-- TODO: remove `twoTorsionPolynomial` in favour of `Ψ₂Sq`
+lemma Ψ₂Sq_eq : W.Ψ₂Sq = W.twoTorsionPolynomial.toPoly :=
+  rfl
+
+/-! ### The univariate polynomials $\tilde{\Psi}_n$ for $n \in \mathbb{N}$ -/
+
+/-- The $3$-division polynomial $\psi_3 = \Psi_3$. -/
+protected noncomputable def Ψ₃ : R[X] :=
+  3 * X ^ 4 + C W.b₂ * X ^ 3 + 3 * C W.b₄ * X ^ 2 + 3 * C W.b₆ * X + C W.b₈
+
+/-- The univariate polynomial $\tilde{\Psi}_4$, which is auxiliary to the $4$-division polynomial
+$\psi_4 = \Psi_4 = \tilde{\Psi}_4\psi_2$. -/
+protected noncomputable def Ψ₄' : R[X] :=
```

Comment:
Do you agree with my suggestion?
```suggestion
protected noncomputable def preΨ₄ : R[X] :=
```

Reply:
Whoops I missed this. I added it on the train but ran out of internet.